File: codeforces/1305D.py
'''
	CodeForces 1305D, Ozon Tech Challenge 2020
	Kuroni and the Celebration

	Tags: Ad-hoc, Simulation
'''
from sys import stdout
from itertools import chain, groupby
from collections import Counter

gkey = lambda p: p[0]

def delNode(node, adj, leaves):
	target = next(iter(adj[node]))
	adj[node].remove(target)
	adj[target].remove(node)
	if len(adj[target]) == 1:
		leaves.append(target)

def solve(edges):
	groups = groupby(sorted(chain(*map(lambda p: [p, (p[1], p[0])], edges)), key=gkey), gkey)
	adj    = dict(map(lambda ps: (ps[0], set(map(lambda p: p[1], ps[1]))), groups))
	leaves = [ k for k, v in adj.items() if len(v) == 1 ]
	while len(leaves) > 1:
		u = leaves.pop()
		v = leaves.pop()
		print(f'? {u} {v}')
		stdout.flush()
		ans = input()
		if ans == u or ans == v:
			return ans
		delNode(u, adj, leaves)
		delNode(v, adj, leaves)
	return leaves[0]

n = int(input())
edges = [ tuple(input().split()) for _ in range(n - 1) ]

print('! {}'.format(solve(edges)))

File: since2020/CodeForces/1305D.cpp
/**
 *  @judge CodeForces
 *  @id 1305D
 *  @name Kuroni and the Celebration
 *  @contest Ozon Tech Challenge 2020
 * 
 *  @tag Ad-hoc, Simulation
 */
#include <iostream>
#include <cstdio>
#include <vector>
using namespace std;

#define MAX 1010

int n, remain[MAX];
vector<int> edges[MAX], leaves;

void del_node(int x) {
	int y;
	for (int i = 0; i < edges[x].size(); i++) {
		y = edges[x][i];
		if (remain[y]) {
			remain[x]--;
			remain[y]--;
			if (remain[y] == 1)
				leaves.push_back(y);
			return ;
		}
	}
	return ;
}

int solve() {
	leaves.clear();
	for (int i = 1; i <= n; i++) {
		remain[i] = edges[i].size();
		if (remain[i] == 1)
			leaves.push_back(i);
	}
	int u, v, ans;
	while (leaves.size() > 1) {
		u = leaves[0];
		v = leaves[1];
		leaves.erase(leaves.begin());
		leaves.erase(leaves.begin());
		printf("? %d %d\n", u, v);
		fflush(stdout);
		cin >> ans;
		if (ans == u || ans == v)
			return ans;
		del_node(u);
		del_node(v);
	}
	return leaves[0];
}

int main() {
	int u, v;
	cin >> n;
	for (int i = 0; i < n - 1; i++) {
		cin >> u >> v;
		edges[u].push_back(v);
		edges[v].push_back(u);
	}
	int ans = solve();
	printf("! %d\n", ans);
}